### DISCUSSION TITLE
Ability to create different profiles.

### DISCUSSION THREAD

**Discussion by eznix86** [zed-industries/zed#23958] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Currently we have only 1 _default_ setting. The problem arises for people working on different projects which needs different settings. 

Example; We can have Rust Profile (which has all the settings for the best experience for Rust). Laravel Profile, Django Profile. Python profile etc. It doesn't need to be specific. But we just need to have the ability to have different profiles.

### If applicable, add mockups / screenshots to help present your vision of the feature

We can have prebuilt profiles (for specific languages/frameworks) or we can have custom profiles. Those profiles can be exported, imported, deleted, extended.

**Comment by jansol** [comment]
IIRC there are already some per-language settings, but there definitely is more that could be done with that.

**Comment by JosephTLyons** [comment]
We also have local settings as well - which can kind of act as a bandaid in this case, though, it isn't ideal in the context of the  original request.

**Comment by ceopaludetto** [comment]
Personally, I think this should follow something like `tsconfig.json` does. So I want to introduce two new concepts:

1. The `extends: string | string[]` + `name: string` options, this will allow us to do:

```jsonc
{
  // The base profile
  "name": "base",

  "theme": "Catppuccin Mocha",
  "vim_mode": true,
}
```

And then:
```jsonc
{
  // The typescript profile
  "extends": ["base", ...],
  "name": "typescript",

  "languages": {
    "TypeScript": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true
      }
    }
  }
}
```

Primitive options like `string | number | boolean` should be overwritten and `object | array` should be merged together

2. The `extensions: string[]` (or maybe `extensions: { include: string[], exclude: string[] }`) option. We need to disable/enable some extensions in different profiles. Nowadays VSCode allow us to disable in UI, but personally, I prefer to disable via `settings.json`

```jsonc
{
  // The base profile
  "name": "base",

  "theme": "Catppuccin Mocha",
  "vim_mode": true,
}
```

```jsonc
{
  // The swift profile
  "extends": ["base", ...],
  "name": "swift",
  
  "extensions": ["zed:sourcekit-lsp-extension"]
}
```

Some people also like the ability to disable all extensions in the base profile and only enable on specific ones, I think all this possibilities should be supported.

What you think?

**Comment by jansol** [comment]
Hmm... Should the profile also specify the github account and e.g. git author info to use (once we have git commit support)? I.e. separate "work" and "personal" profiles.

Formatting etc options should really be done via local settings IMHO, since different repos have different policies for that.

**Comment by shishkin** [comment]
Local settings work really well for a lot of use cases. However, since local settings are (usually) checked into source control, they are bad for personalization settings. E.g. setting fonts/themes for the whole team is not a good idea. Profiles can solve for that.

But the most need for profiles (as implemented in [VS Code](https://code.visualstudio.com/docs/editor/profiles)) I have is to selectively load extensions based on programming language. While performance concern might not (yet) be an issue for zed, but some extensions can start conflicting with other ones. Again, putting this in local settings may solve part of the problem, but I would hesitate to force all my non-essential extensions on the whole team.

**Reply by olejorgenb** [reply]
> However, since local settings are (usually) checked into source control,

This is why there should always be multiple setting layers IMO. For tools which does not support this I think it's better to provide a template settings file which the developers can choose to merge manually with their own settings file. (or just symlink if they want to use the default)

**Comment by eznix86** [comment]
Hey, thanks for keeping this issue up.

Some feedback on my side:

I believe a profile is per project. we can have a .zed folder which takes the users configuration and merge it with the settings.json living inside of the .zed folder.

This allow teams to share extensions for example. I believe the .zed folder should allow a subset of the configuration, which means, you cannot change theme in a .zed folder.

Basically I believe for sanity, the word "Extending" is just a cloning of the json which technically another profile being duplicated.

So we can have settings.json, and even nvim.json, python.json or more.

**Comment by shishkin** [comment]
> I believe a profile is per project.

That's something I strongly disagree. Profile is a user-specific and idiosyncratic set of extensions and preferences that are often used together. E.g. while working on devops automation I may prefer different layout from that when I work on mobile or web apps. And those preferences are not something I would impose on a team working on the same project. In addition, I would like to easily reuse such profiles for many projects, so initial project setup doesn't involve copying and tweaking zed config from a previous project. Again, project settings are shared, while profiles should be private.

**Comment by eznix86** [comment]
@shishkin I agree, but I wasn't explicit enough, what i meant was a profile can be reused by a project.

Example; you have 10 projects, you have 3 profile. But you open the 9th project in profile number 2, it should reopen with the same profile each time, unless you change it.

But we can have a profile per project. Example; 

Let's say you have your current profile activated, since it has a .zed with settings.json inside it will merge your config with what we have in the .zed folder.

So it is user specific at its core, but if you open different projects on different profile, we should be able to memorize that this project was on which profile.

And in case the project has a project specific .zed/settings.json it should load and merge with your current profile but this project specific settings should not allow everything. Example, it should not override themes, or other user specific configuration.

I hope it clarified or even improved what i said earlier.

**Comment by seg4lt** [comment]
Maybe zed can take some inspiration from VSCode profile feature
https://code.visualstudio.com/docs/editor/profiles

**Comment by cole-miller** [comment]
Thanks everyone for the feedback. I'm going to convert this to a discussion since that's a better venue for hashing out whether we want a feature along these lines and if so what the design should be.

**Comment by mattsays** [comment]
Any updates on this?

**Reply by Anifyuli** [reply]
I waiting this update too

**Comment by isFakeAccount** [comment]
Profile feature is def one of the best features VSCode offers and it would be pretty great to have it on Zed as well.

**Comment by phromo** [comment]
My main usage for profiles would be to have an assistant-only mode where I've maximised the size of the agent panel and I use it standalone for general Q&A. Would be nice to be able to start that with `zed --profile my-agentic` etc.

**Comment by adreasnow** [comment]
I can't see it documented anywhere but in Zed there's the "switch profiles" command in the command palette, and also if you click on your account picture in the top-right of thew window (`ctrl`+`opt`+`cmd`+`P`).

This seems to correlate in `settings.zed` to a block like this:
```json
"profiles": {
  "Home": {},
  "Work": {}
}
```

And switching profiles seems to apply the settings in those blocks 🙂

**Reply by chidniu** [reply]
Is there a way to make one of these profiles the default?

**Reply by rosavalenzuela23** [reply]
Just asking, switching between those profiles also unloads/loads specific extensions?

**Reply by exil0867** [reply]
> Just asking, switching between those profiles also unloads/loads specific extensions?
> 

Did you find an answer?

**Comment by Icehunter** [comment]
I currently maintain two isolated configurations for Cursor—one for personal work and one for corporate work. This lets me keep API keys, authentication tokens, logins, prompts, and settings completely separate.

```bash
_launch_cursor_app() {
    local user_dir="$1"
    shift

    if [[ $# -eq 0 ]]; then
        command cursor --user-data-dir="$user_dir" .
    else
        command cursor --user-data-dir="$user_dir" "$@"
    fi
}

# Personal profile (default "cursor")
cursor() {
  _launch_cursor_app "$HOME/Library/Application Support/Cursor" "$@"
}

# Work profile (short alias "cw")
cw() {
  _launch_cursor_app "$HOME/Library/Application Support/Cursor-Work" "$@"
}
```

This gives me:

cursor → personal environment
cw → corporate environment

Zed does support an override of the user-data directory, but currently:

- Conversations, prompts, API keys, and various settings are stored inside the Zed config directory
- (e.g., $HOME/.config/zed)
- The user-data directory alone is not enough to separate these kinds of sensitive or identity-specific assets

This makes it difficult to maintain clean personal vs. corporate profiles. Even with a separate user-data directory, both profiles still end up sharing:

- Model/API keys
- Prompt history and conversations
- Editor settings
- AI behavior configuration
- Other identity-linked data

Ideally, Zed would allow something like:

- Personal config dir → $HOME/.config/zed
- Corporate config dir → $HOME/.config/zed-corp

Or a flag/env-var such as:

```bash
zed --config-dir=/path/to/alternate/zed-config
ZED_CONFIG_DIR=/path/to/alternate/zed-config zed
```

Is this possible now?

If not, is support for selecting or overriding the config directory on the roadmap? 

Thanks!

**Comment by toihr** [comment]
I just started using Zed on Windows. Because of the great performance. Now with extensions don't I have all extensions in one profile and would this not slow down my zed, like Vscode? Because even if I write JS it also loads my .NET, Rust etc. extensions?

**Comment by exil0867** [comment]
If you use [Devenv](https://devenv.sh/) you can do something like:
```shell
#!/usr/bin/env bash
set -euo pipefail

PROJECT="project name"

export XDG_CONFIG_HOME="$HOME/.config/zed-$PROJECT"
export XDG_DATA_HOME="$HOME/.local/share/zed-$PROJECT"

exec devenv shell -- zeditor .

```
<img width="2560" height="1440" alt="Screenshot From 2026-01-06 02-19-02" src="https://github.com/user-attachments/assets/9e6e5fbb-0435-4dba-9423-435c6e5bb584" />

**Comment by rsmck** [comment]
This partially addresses the concern I raised in https://github.com/zed-industries/zed/discussions/40472 

However, it seems that the profiles function still doesn't allow you to set different agent parameters unless I'm doing something wrong.  

I have multiple sets of AI credentials, some are for my own AI subscriptions and others are, for example, Github Enterprise for a specific company - it would be really useful to switch between them.  I have found that the following; 

`  "profiles": {
    "Acme Inc": {
      "disable_ai": false,
    },
    "Personal": {
      "disable_ai": true,
    },
  },
`

Allows me to at least _disable_ the AI functionality by switching to another profile, but putting the `edit_predictions` and `agent` blocks in here doesn't seem to work?

**Comment by sultanmyrza** [comment]
I always use command panel for quick switch and hope zed can have similar UX

<img width="1512" height="982" alt="Screenshot 2026-02-02 at 2 12 14 PM" src="https://github.com/user-attachments/assets/d4d42d86-d9da-4490-afc9-ba82458adcb6" />

<img width="1512" height="982" alt="Screenshot 2026-02-02 at 2 12 19 PM" src="https://github.com/user-attachments/assets/2184d183-b759-491f-b60d-273e1fd8037a" />

**Comment by toihr** [comment]
I want to maybe add a different spin on this. For me the question about profiles is about not loading all plugins all the time. Maybe if we could define some kind of lazy plugin load. Aka. only load the Rust plugin if an .rs file is opened. 

For non language specific plugins, maybe we can define it on a per project basis.

**Comment by c10l** [comment]
I'd love to be able to run fully isolated Zed instances, for different clients and personal projects.

One of the main reasons is that clients have different policies and offerings for agents and LLMs, and I have a separate configuration for my personal projects. Not being able to sandbox their use for any given Zed profile/project/window is a no-go.

**Comment by daveycodez** [comment]
I need to be able to swap profiles on the fly, so I can easily switch between fully offline LLM mode and online LLM mode. In other words, I want to have one profile that uses Zed Agent with Claude Opus & Edit Predictions with Zeta and another `offline` Profile that uses Zed Agent with LM Studio models & also uses Edit Predictions with my local model.

Another option would be to add some way to easily switch the model for edit predictions without having to edit the settings.json each time.

**Reply by JonGretar** [reply]
And something like this would not suffice?

```JSONC
"profiles": {
    "Offline Mode": {
      "base": "user",
      "settings": {
        "theme": "Catppuccin Macchiato",
        "agent": {
          "default_model": {
            "model": "kimi-k2.5",
            "provider": "Ollama",
          },
        },
      },
    },
  },
```
